prototypes/matrix_based_non_linear_poisson/matrix_based_non_linear_poisson.cc

```diff
@@ -0,0 +1,1486 @@
+                                            phi_i * nonlinearity * phi_j) *
+                                           dx;
+                    }
+                }
+            }
+
+          cell->get_dof_indices(local_dof_indices);
+          constraints.distribute_local_to_global(cell_matrix,
+                                                 local_dof_indices,
+                                                 system_matrix);
+        }
+    }
+
+  system_matrix.compress(VectorOperation::add);
+}
+
+template <int dim, int fe_degree>
+void
+MatrixBasedPoissonProblem<dim, fe_degree>::assemble_gmg()
+{
+  QGauss<dim> quadrature_formula(fe_degree + 1);
+
+  FEValues<dim> fe_values(fe,
+                          quadrature_formula,
+                          update_values | update_gradients |
+                            update_quadrature_points | update_JxW_values);
+
+  const unsigned int dofs_per_cell = fe_values.dofs_per_cell;
+  const unsigned int n_q_points    = fe_values.n_quadrature_points;
+
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+
+  std::vector<double> newton_step_values(n_q_points);
+
+  std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+  std::vector<AffineConstraints<double>> boundary_constraints(
+    triangulation.n_global_levels());
+
+  for (unsigned int level = 0; level < triangulation.n_global_levels(); ++level)
+    {
+      const IndexSet dof_set =
+        DoFTools::extract_locally_relevant_level_dofs(dof_handler, level);
+      boundary_constraints[level].reinit(dof_set);
+      boundary_constraints[level].add_lines(
+        mg_constrained_dofs.get_refinement_edge_indices(level));
+      boundary_constraints[level].add_lines(
+        mg_constrained_dofs.get_boundary_indices(level));
+
+      boundary_constraints[level].close();
+    }
+
+  for (const auto &cell : dof_handler.cell_iterators())
+    if (cell->level_subdomain_id() == triangulation.locally_owned_subdomain())
+      {
+        cell_matrix = 0;
+
+        fe_values.reinit(cell);
+
+        fe_values.get_function_values(solution, newton_step_values);
+
+        for (unsigned int q = 0; q < n_q_points; ++q)
+          {
+            const double nonlinearity = std::exp(newton_step_values[q]);
+            const double dx           = fe_values.JxW(q);
+
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+              {
+                const double         phi_i      = fe_values.shape_value(i, q);
+                const Tensor<1, dim> grad_phi_i = fe_values.shape_grad(i, q);
+
+                for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                  {
+                    const double         phi_j = fe_values.shape_value(j, q);
+                    const Tensor<1, dim> grad_phi_j =
+                      fe_values.shape_grad(j, q);
+
+                    cell_matrix(i, j) +=
+                      (grad_phi_i * grad_phi_j - phi_i * nonlinearity * phi_j) *
+                      dx;
```

Comment:
No need to do this here, but there's a hack you could do to only multiply grad_phi_i and grad_phi_j when component_i==component_J. No need to do it, just FYI.